### ISSUE TITLE
Export public symbols to WebAssembly by default

### ISSUE THREAD

**Issue by Meai1** [jondgoodwin/cone#10]
As we've discussed on irc, it is possible to export them all through:

`conec main.cone --wasm && wasm-ld ./main.wasm -o main.wasm --no-entry --import-memory --allow-undefined-file=libc.imports --export-all`

I check whether the output has generated exported symbols by using the wasm2wat tool, compiled from here: https://github.com/WebAssembly/wabt

But the idea is to export only public functions / globals, not everything - so that's more like a hack.

I did some research on this, this is probably the best link that explains it: 
https://lld.llvm.org/WebAssembly.html#imports-and-exports

> When building a shared library any symbols marked as visibility=default will be exported. When building an executable, only the entry point and symbols flagged as WASM_SYMBOL_EXPORTED are exported by default. In LLVM the WASM_SYMBOL_EXPORTED flag is applied to any symbol in the llvm.used list which corresponds to __attribute__((used)) in C/C++ sources.
> In addition, symbols can be exported via the linker command line using --export.
> Finally, just like with native ELF linker the --export-dynamic flag can be used to export symbol in the executable which are marked as visibility=default.

LLVM LangRef.rst states the particulars on how to do it (but only in LLVM IL unfortunately)

```
The ``@llvm.used`` global is an array which has
:ref:`appending linkage <linkage_appending>`. This array contains a list of
pointers to named global variables, functions and aliases which may optionally
have a pointer cast formed of bitcast or getelementptr. For example, a legal
use of it is:

.. code-block:: llvm

    @X = global i8 4
    @Y = global i32 123

    @llvm.used = appending global [2 x i8*] [
       i8* @X,
       i8* bitcast (i32* @Y to i8*)
    ], section "llvm.metadata"

If a symbol appears in the ``@llvm.used`` list, then the compiler, assembler,
and linker are required to treat the symbol as if there is a reference to the
symbol that it cannot see (which is why they have to be named). For example, if
a variable has internal linkage and no references other than that from the
``@llvm.used`` list, it cannot be deleted. This is commonly used to represent
references from inline asms and other things the compiler cannot "see", and
corresponds to "``attribute((used))``" in GNU C.
```

This is how far I made it:

```

        LLVMTypeRef arrType = LLVMArrayType(LLVMPointerType(func_type, LLVMGetPointerAddressSpace(func_type)), 1);
        LLVMValueRef arrInstance = LLVMAddGlobal(gen->module, arrType,"llvm.used");
        LLVMSetSection(arrInstance, "llvm.metadata");
        LLVMSetLinkage(arrInstance, LLVMAppendingLinkage);
```
I dont know if this comes close but I think it might. It's just trying to create this global array of pointers that was described in the .rst doc above. I dont know how to use the C API to populate the array with a pointer to the function.

**Comment by jondgoodwin** [comment]
Appreciate you doing that research on how one might build the LLVM IR to mark globals with the equivalent of attribute((used)) in C. I suspect you do have much of it worked out.

That said, I think there is a simpler option available near-term.  Cone does distinguish between public and private globals (the latter when a name begins with an underscore), and it already marks them differently to LLVM:  public globals are marked with default visibility and private ones are not. So, if you use wasm-ld's --export-dynamic, that will only take the public globals and not the private ones.

Let me know if that works for you, as I cannot test it today.

**Comment by Meai1** [comment]
No, --export-dynamic also exports the `fn _bro()` as well as the `fn blah()`
The only option to limit it to certain functions seems to be to specifically pass it via `--export=blah`, then it exports only exactly that one.

**Comment by Meai1** [comment]
ok @jondgoodwin, it does work but only like this because it never jumps into the first if block and even if it does, the x86 calling convention thing ruins it for wasm so I would suggest this:

```
        if (glofn->flags & FlagSystem)
        {
            LLVMSetFunctionCallConv(glofn->llvmvar, LLVMX86StdcallCallConv);
            LLVMSetDLLStorageClass(glofn->llvmvar, LLVMDLLImportStorageClass);
            LLVMSetVisibility(glofn->llvmvar, LLVMHiddenVisibility);
        }
        else if(fnname[0] == '_'){
            LLVMSetVisibility(glofn->llvmvar, LLVMHiddenVisibility);
        }
```

**Comment by jondgoodwin** [comment]
I took a quick pass in committing a code change so that private globals use HiddenVisibility (assuming that if unspecified, the remaining ones will have DefaultVisibility). I did not add the ExternalLinkage, as I am currently unsure that it is needed. 

"extern"s (the first if condition) should not be marked with HiddenVisibility, as that defeats the purpose of being able to find and link them in!

I am unable to test the change right now, but if you are, let me know if it works for you.
